### DISCUSSION TITLE
Telescope PR merged and I have previews again

### DISCUSSION THREAD

**Discussion by dimitarvp** [LunarVim/LunarVim#3876] [Show and tell]
_**(NOT MY WORK)**_

After this NeoVim Telescope PR got merged -- https://github.com/nvim-telescope/telescope.nvim/pull/2395 -- I have put the following in my `config.lua`:

```lua
lvim.builtin.which_key.mappings["f"] = {
  require("lvim.core.telescope.custom-finders").find_project_files,
  "Project files"
}

lvim.builtin.which_key.mappings.b.f = {
  "<cmd>Telescope buffers<cr>",
  "Buffers"
}
```

...and both project files and buffer switcher now show me previews.

Shout out to @LostNeophyte for engaging with me on Reddit, and here's a pointer to my final comment there: https://www.reddit.com/r/lunarvim/comments/1014usu/recent_update_changed_previewers/j96a0a5/

If there's a better way, let me know (still a VIM noob). If not, feel free to ignore, I just wanted to bring visibility to the Telescope team (and @Conni2461 in particular!) that fixed an issue that was a productivity breaker for me. And now LunarVim benefits from that as well! Awesome job from everyone involved in both communities. ❤️

**Comment by LostInTheLogs** [comment]
the code you have in your config it the current(on master) correct way to enable previews

it should work even before that telescope pr

**Comment by dimitarvp** [comment]
You told me that but it didn't, sadly.

I tried calling `:Telescope buffers` and that errored out.  The PR fixed it and after that it started working.

Though thinking of it maybe I have to check all GIT repos inside LunarVim, maybe one of them refused to update. (Though I'd expect `LvimUpdate` to do something like `git reset --hard HEAD` before doing `git pull`.)

**Reply by LostInTheLogs** [reply]
interesting, it works for me
![image](https://user-images.githubusercontent.com/110467150/220152489-92a49d87-ad43-41e0-b745-e0f10912e698.png)

**Reply by dimitarvp** [reply]
This is most of my `config.lua` (the rest is TreeSitter, LSP and Telescope exclude patterns). I also updated LunarVim 2 days ago.

```lua
lvim.builtin.telescope.pickers = {
  buffers = {
    --
    -- Setting theme to "dropdown" makes the buffer selector smaller and centered.
    -- I like the selector like all others -- big and with prompt at the bottom -- so this
    -- will stay commented out for the moment.
    --
    -- theme = "dropdown",
    previewer = true,
    initial_mode = "insert",
    mappings = {
      i = {
        ["<C-d>"] = actions.delete_buffer,
      },
      n = {
        ["dd"] = actions.delete_buffer,
      },
    },
    scroll_strategy = "limit",
  },
}

lvim.builtin.telescope.defaults.scroll_strategy = "limit"
lvim.builtin.telescope.defaults.layout_config = {
  -- prompt_position = "top",
  height = 0.95,
  width = 0.95,
  bottom_pane = {
    height = 25,
    preview_cutoff = 120,
  },
  center = {
    height = 0.4,
    preview_cutoff = 40,
    width = 0.5,
  },
  cursor = {
    preview_cutoff = 40,
  },
  horizontal = {
    preview_cutoff = 120,
    preview_width = 0.6,
  },
  vertical = {
    preview_cutoff = 40,
  },
  flex = {
    flip_columns = 150,
  },
}

lvim.builtin.telescope.theme = "center"
lvim.builtin.telescope.defaults.initial_mode = "insert"
-- lvim.builtin.telescope.defaults.sorting_strategy = "ascending"
lvim.builtin.telescope.defaults.layout_strategy = "flex"
lvim.builtin.telescope.defaults.prompt_prefix = "  "
lvim.builtin.telescope.defaults.selection_caret = "❯ "
lvim.builtin.telescope.defaults.mappings.i["<esc>"] = actions.close
lvim.builtin.telescope.defaults.winblend = 10

--
-- This gives us previewing of files when looking through project files.
-- Without this we only get a list of files without a preview.
-- TODO: find a better way! Likely directly assigning pickers.
--
lvim.builtin.which_key.mappings["f"] = {
  require("lvim.core.telescope.custom-finders").find_project_files,
  "Project files"
}

-- Show previewer when searching buffers with <leader>bf
lvim.builtin.which_key.mappings.b.f = {
  "<cmd>Telescope buffers<cr>",
  "Buffers"
}

-- After changing plugin config exit and reopen LunarVim, Run :PackerInstall :PackerCompile
lvim.builtin.alpha.active = true
lvim.builtin.alpha.mode = "dashboard"
lvim.builtin.terminal.active = true
lvim.builtin.nvimtree.setup.view.side = "left"
lvim.builtin.nvimtree.setup.renderer.icons.show.git = false
```

**Reply by LostInTheLogs** [reply]
your config works as well, assuming you did require telescope's actions somewhere above this code
![image](https://user-images.githubusercontent.com/110467150/220154212-483d41d4-adc4-40f5-80cb-94f846e11ea4.png)

**Reply by dimitarvp** [reply]
So we're back to square one. I'll do another clean reinstall... sometime in the future. 😅

Is there a _complete_ script used to clean everything and then reinstall? I devised my own a while ago:

```sh
rm ~/.local/bin/lvim
rm -rf ~/.local/share/lunarvim
rm -rf ~/.config/lvim
rm -rf ~/.cache/lvim
```

Is that enough to do before running the installer?

**Reply by LostInTheLogs** [reply]
the installer cleans all the things, you can run it without uninstalling, but we also have an uninstaller

**Reply by dimitarvp** [reply]
I'll backup my `config.lua` of course -- but does the installer wipe it while it does its thing?

**Reply by LostInTheLogs** [reply]
it moves it to ~/lvim.old
but it's always safer to back it up anyways

**Reply by dimitarvp** [reply]
Right now I want to install the stable release -- the `LV_BRANCH='release-1.2/neovim-0.8'` prefix when running the installer -- but what happens when NeoVim stable moves to 0.9 and LunarVim moves to 1.3.0 and beyond? Does `LvimUpdate` take care of moving to next release after it's out?

**Reply by LostInTheLogs** [reply]
no, LvimUpdate doesn't change the branches, most such updates require a reinstall anyway

**Reply by dimitarvp** [reply]
Just reinstalled and out of the box I only get a center (the `dropdown` theme) list of buffers. No preview. Haven't put anything from my `config.lua` yet.

Then I reapplied my `config.lua` and I still get the Lua error shown in the Telescope PR. No idea man, fresh reinstall.

**Reply by dimitarvp** [reply]
What's even funnier: I can see the change in the Telescope PR -- it's in `~/.local/share/lunarvim.old/site/pack/lazy/opt/telescope.nvim/lua/telescope/pickers.lua`, lines 1423-1424.

**Reply by LostInTheLogs** [reply]
what if you remove the `previewer = true` line

**Reply by dimitarvp** [reply]
Just tried. Still the same error.

BTW the file finder works fine and shows previews. The problems are only with the buffer switcher.

**Reply by dimitarvp** [reply]
WAIT! My mistake. Updated the old file, not the new one.

Now it works. So the `previewer = true` is tripping it up, interesting.

**Reply by LostInTheLogs** [reply]
well, idk what's wrong, it works after the pr and that's all that matters :)

**Reply by LostInTheLogs** [reply]
> WAIT! My mistake. Updated the old file, not the new one.
> 
> Now it works. So the `previewer = true` is tripping it up, interesting.

telescope didn't allow true as an option before, you needed to set it to  some specific previewer

**Reply by dimitarvp** [reply]
Yeah but now it seems it should allow it?

Nevermind though, as you said I have a fully working setup + an updated LunarVim so win all around. :)

I have to say for the record that I absolutely hate such weirdnesses -- and that's why people go to VSCode and other IDEs. But I love the command line too much so I'll keep tolerating these problems and troubleshoot them when possible.

Again, thanks for your work! LunarVim is awesome and I love it.

**Reply by LostInTheLogs** [reply]
> Yeah but now it seems it should allow it?

after the pr, yes

> I have to say for the record that I absolutely hate such weirdnesses

yea, I have telescope's config structure too, it's quite hard to understand

> Again, thanks for your work! LunarVim is awesome and I love it.

no problem :) if you need any help in the future i recommend Chris' discord server over reddit, I'm there on the #lunarvim channel most of the time and very rarely i take a look at reddit

**Reply by Conni2461** [reply]
I think we did quite well with the telescope config structure with the `defaults` and `pickers` table. But i have to admit, its a huge issue that we have `preview` and `previewer` that do different things and accept different values. Its also an ossue that there are a tone of internal opts that can be overriden (like `previewer` and `sorter`, ...) and those arnt even documented. So yeah a couple of bad decisions are a huge issue and can't be addressed without breaking changes. And every time i do a documented breaking change i have to deal with 20 issues over the next couple of months ....

**Reply by Conni2461** [reply]
Thats also the reason why i just implemented support for `previewer = true`. Its easier than to explain everything ... and now it actually makes sense. `previewer` can be a bool, fn pointer to a previewer or a table of fn pointers to previewers. Rather than only accepting false, fn pointers ...

**Reply by Conni2461** [reply]
We could also just introduce strong typing if that would resolve those kind of issues. But that would break a bunch of configs.

**Reply by dimitarvp** [reply]
If we want to follow semantic versioning, Neovim isn't at 1.0 yet. Maybe you can plan such breaking changes for then?

Personally I'd always vote for strong typing. I love dynamic languages and they made me loads of money but there comes one point after which it's a hassle when stuff is so... free and wild, let's call it.

Though as a guy who had to maintain a lot of old software I believe your users -- myself included -- will be most receptive to breaking changes if Neovim itself runs a migration script to change the old configs to the new ones (if such a thing is even possible; I suspect it's not).

**Reply by Conni2461** [reply]
Uff migration script will never happen. That would require us to write inside your lua config files. Thr best we can do, and we've done that in the past, is offer a migration guide once we realize that your config is out of date.

We, telescope, is also still on 0.1 for this reason. We dont guarantee anything. We released version 0.1.0 and have a stable release branch 0.1.x because we want user away from master, which is sadly the default for all plugin managers. So we can finally expermient again on master, which is currently basically impossible because every experiment, will be instantly roll out to the majority of users and i hate it. Luarocks could also solve this issue because it installs the latest stable version not master but packer.nvim is the only plugin manager with support for that anf with the move to lazy.nvim i dont see that happening.

I am happy that lunarvim adopted to using 0.1.x :) but i am sure a lot of ppl are still running current master

**Reply by dimitarvp** [reply]
Well okay then. I'd love it if at least Telescope can show a warning akin to "your config seems out of date, please visit this website to migrate"? That can help.

**Reply by Conni2461** [reply]
we did show a error that points to `:help telescope.changelog` but yaeh

**Reply by LostInTheLogs** [reply]
> I think we did quite well with the telescope config structure with the defaults and pickers table. But i have to admit, its a huge issue that we have preview and previewer that do different things and accept different values. Its also an ossue that there are a tone of internal opts that can be overriden (like previewer and sorter, ...) and those arnt even documented. So yeah a couple of bad decisions are a huge issue and can't be addressed without breaking changes. And every time i do a documented breaking change i have to deal with 20 issues over the next couple of months ....

The config is good when you learn what's what, but it took me a long time to learn it (compared to most plugins), doc/telescope.txt is very long and there are a lot of options. telescope is just extremely configurable, it's overwhelming. the part that took me the longest is is `theme` in `config.pickers` table (like [here](https://github.com/LunarVim/LunarVim/blob/36c8bdee9ff59a0a63c1edfc445b5eb2886cf246/lua/lvim/core/telescope.lua#L37)) it took me the most time and I can't find any documentation for it, I probably should have opened a discussion on the repo to ask about it. [this part of the docs](https://github.com/nvim-telescope/telescope.nvim/blob/3915d933dc19d31571d2953ca9461fe6fa117a6e/doc/telescope.txt#L86-L108) only shows options for `defaults` and doesn't link to anything regarding `pickers` or the explanation is too short for me, it doesn't explicitly say what are the allowed keys, it could say they are the same exact keys from `builtin.<picker_name>({opts})`, but that doesn't seem to be the case as i don't see `theme` there. If that part of the docs explained it I think I wouldn't have any problem with the config

I can relate to the documented breaking changes thing, even if you pin an issue with a solution and the error/warning links to it people will still open issues

apart from that one thing I love the plugin, please don't take my criticism harshly and thank you very much for working on telescope :heart:

**Comment by Conni2461** [comment]
> doc/telescope.txt is very long and there are a lot of options. telescope is just extremely configurable, it's overwhelming

i agree, the problem, imo is that telescope is too big. so we end up with a huge documentation that doesnt even include everything

> the part that took me the longest is is theme

yep theme was a good idea with a bad architecture. the problem is that theme is a high level wrapper around layout_config with is just a lower level config for layout_strategy which are also just recipes on how windows should be drawn. Resulting in multiple ways of configuring the same thing.

> If that part of the docs explained it I think I wouldn't have any problem with the config

yep fair enough. PR welcome :)
